File: backend/lib/gnuplex-display-server/litedb.rb
require "fileutils"
require "sqlite3"
require_relative "mpv_cmd"

class LiteDB
  def initialize
    db.execute <<-SQL
      create table if not exists pos_cache (
        filepath string not null primary key,
        pos int
      );
    SQL
    db.execute <<-SQL
      create table if not exists history (
	      id	integer not null unique,
      	mediafile	text,
	      primary key("id" AUTOINCREMENT)
      );
    SQL
    db.execute <<~SQL
           create table if not exists "medialist" (
          	"filepath"	text not null,
      	    primary key("filepath")
      ) 
    SQL
  end

  def db
    FileUtils.mkdir_p File.join(File.dirname(__FILE__), "..", "..", "..", "tmp")
    filepath = File.join(File.dirname(__FILE__), "..", "..", "..", "tmp", "gnuplex.sqlite3")
    @db ||= SQLite3::Database.new filepath
  end

  def first_or_default(query, args, default)
    res = db.execute query, args
    res.flatten[0] || default
  end

  def read_query(query, args)
    res = []
    db.execute(query) do |row|
      res.append row
    end
    res.flatten
  end

  def loadpos(filepath)
    first_or_default <<-SQL, [filepath]
      select pos from pos_cache where filepath = ?;
    SQL
  end

  def savepos(filepath, pos)
    db.execute <<-SQL, [filepath, pos]
      insert or replace into pos_cache (filepath, pos) values (?, ?);
    SQL
  end

  def medialist
    read_query <<-SQL, []
      select distinct filepath from medialist order by filepath;
    SQL
  end

  def refresh_medialist(medialist)
    medialist.each do |filepath|
      db.execute <<-SQL, [filepath]
        insert or replace into medialist (filepath) values (?);
      SQL
    end
  end

  def addhist(filepath)
    db.execute <<-SQL, [filepath]
      insert into history (mediafile) values (?);
    SQL
  end

  def last25
    read_query <<-SQL, []
      select distinct mediafile from history order by id desc limit 25;
    SQL
  end
end

File: backend/liteDB/liteDB.go
package liteDB

import (
	"database/sql"
	"fmt"
	"gnuplex-backend/consts"
	"log"
	"strconv"
	"sync"

	_ "modernc.org/sqlite"
)

type LiteDB struct {
	SqliteConn *sql.DB
	Mu         *sync.Mutex
}

func Init() *LiteDB {
	var db LiteDB
	db.Mu = &sync.Mutex{}
	db.Mu.Lock()
	fmt.Println("Got Init LiteDB lock")
	defer db.Mu.Unlock()
	defer fmt.Println("Rem Init LiteDB lock")
	conn, err := sql.Open("sqlite", "../tmp/gnuplex.sqlite3")
	db.SqliteConn = conn
	if err != nil {
		log.Fatal("Init LiteDB fatal error:", err)
	}
	_, err = db.SqliteConn.Exec("create table if not exists pos_cache (filepath string not null primary key, pos int);")
	if err != nil {
		log.Fatal("Init LiteDB error 1:", err)
	}
	_, err = db.SqliteConn.Exec("create table if not exists history (id integer not null unique, mediafile	text, primary key(id AUTOINCREMENT));")
	if err != nil {
		log.Fatal("Init LiteDB error 2:", err)
	}
	_, err = db.SqliteConn.Exec("create table if not exists medialist (filepath text not null,  primary key(filepath)) ;")
	if err != nil {
		log.Fatal("Init LiteDB error 3:", err)
	}
	_, err = db.SqliteConn.Exec("create table if not exists mediadirs (filepath text not null, primary key(filepath)) ;")
	if err != nil {
		log.Fatal("Init LiteDB error 4:", err)
	}
	_, err = db.SqliteConn.Exec("create table if not exists version_info (key string not null primary key, value string);")
	if err != nil {
		log.Fatal("Init LiteDB error 5:", err)
	}
	_, err = db.SqliteConn.Exec("insert or ignore into version_info values ('db_schema_version', ?);", consts.DBVersion)
	if err != nil {
		log.Fatal("Init LiteDB error 6:", err)
	}
	UpgradeDB(&db)
	return &db
}

func UpgradeDB(liteDB *LiteDB) {
	rows, err := liteDB.SqliteConn.Query("select value from version_info where key = 'db_schema_version';")
	if err != nil {
		log.Fatal("Upgrade db error:", err)
	}
	next := rows.Next()
	if !next {
		log.Fatal("Upgrade db error: no version schema", err)
	}
	var vers string
	err = rows.Scan(&vers)
	if err != nil {
		log.Fatal("Bad version schema 1", err)
	}
	versNum, err := strconv.Atoi(vers)
	if err != nil {
		log.Fatal("Bad version schema 2", err)
	}
	if versNum < consts.DBVersion {
		fmt.Println("Should I do something?")
	}
	rows.Close()
}
